subsys/portability/cmsis_rtos_v1/cmsis_thread.c

```diff
@@ -164,6 +175,20 @@ osStatus osThreadTerminate(osThreadId thread_id)
 	}
 
 	k_thread_abort((k_tid_t)thread_id);
+
+	/* get instance index according to thread stack start address */
+	osThreadDef_t *thread_def =
+		(osThreadDef_t *)(k_thread_custom_data_get());
```

Comment:
Custom data can be set outside of CMSIS via `k_thread_custom_data_set()`, which is a public API. Once that function is called, the `thread_def` pointer is no longer valid. So please find another way of retrieving this data. Also, please remove the `k_thread_custom_data_set()` above.

Reply:
@dcpleung  yes, thanks for your point, I agree with you, we can't use custom_data of thread as the reference to terminate threads, because it will be easily changed and cause some issues. . i.e the osThreadTerminate(id1) doesn't work as expected in the code below, it tries to free id1 in the os_thread_def_thread2, this is incorrect.

> id1 = osThreadCreate(osThread(thread1), NULL);
> id2 = osThreadCreate(osThread(thread2), NULL);
> osThreadTerminate(id1)

but according to cmsis rtos doc, the osThreadTerminate API shall be consistent, so if I want to free a thread instance just by struct k_thread, the easiest way is to ****add a new configurable pointer variable into struct k_thread** using config like if enabling CONFIG_CMSIS_RTOS_V1, and make this variable point to it's source os_thread_def struct**, this method should work, but I am not sure whether this is an appropriate way, could you give some advice on this? thanks.

and for the k_thread_custom_data_set(), it was existed in osThreadCreate function of original code, because this was designed to find source os_thread_def of current thread in osSignalWait.

Reply:
Hm... let's focus on getting the instance counter working first. For now, keep using custom data for now and file an enhancement issue on fixing it.

Reply:
@dcpleung  as I understand, I can release thread instance resource in osThreadTerminate for terminating those active threads, but threads can also be terminated by just returning from its thread function, then where should I release instance for them?

Reply:
I tried to add thread instance release operation into low level k_thread_abort function. because no matter whether a thread was terminated by other threads or returning from its thread function, they all should release its thread instance.

Reply:
Hm... I think that's probably a good place to do clean up. Can you also add some tests for this scenairo?

Reply:
@dcpleung yes, I added a test in existing "tests/subsys/portability/cmsis_rtos_v1/src/thread_instance.c" to verify this.
terminate a thread after thread instances get to its limit, then create a new thread and check whether the new thread could be created successfully and whether the thread instance address is the same as the terminated thread.